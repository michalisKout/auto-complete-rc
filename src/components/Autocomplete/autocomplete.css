.autocomplete {
  position: relative;
  width: 100%;
  max-width: 400px;
}

.autocomplete__input {
  width: 100%;
  padding: 12px 16px;
  border: 2px solid var(--bg-color);
  border-radius: 8px;
  font-size: 16px;
  line-height: 1.5;
  background-color: white;
  transition: border-color 0.2s ease-in-out, box-shadow 0.2s ease-in-out;
  outline: none;
  box-sizing: border-box;
}

.autocomplete__input:focus {
  border-color: var(--primary-color);
}

.autocomplete__input:disabled {
  background-color: var(--bg-color);
  color: var(--fallback-text);
  cursor: not-allowed;
}
