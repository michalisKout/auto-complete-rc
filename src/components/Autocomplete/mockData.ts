import type { AutocompleteItem } from "./types";

/**
 * Mock data for autocomplete component
 * Simulates a realistic dataset that might come from an API
 * (** Generated with the use of AI tools)
 */
export const mockData: AutocompleteItem[] = [
  { id: 1, label: "Apple", value: "apple" },
  { id: 2, label: "Banana", value: "banana" },
  { id: 3, label: "Cherry", value: "cherry" },
  { id: 4, label: "Date", value: "date" },
  { id: 5, label: "Elderberry", value: "elderberry" },
  { id: 6, label: "Fig", value: "fig" },
  { id: 7, label: "Grape", value: "grape" },
  { id: 8, label: "Honeydew", value: "honeydew" },
  { id: 9, label: "Kiwi", value: "kiwi" },
  { id: 10, label: "Lemon", value: "lemon" },
  { id: 11, label: "Mango", value: "mango" },
  { id: 12, label: "Orange", value: "orange" },
  { id: 13, label: "Papaya", value: "papaya" },
  { id: 14, label: "Quince", value: "quince" },
  { id: 15, label: "Raspberry", value: "raspberry" },
  { id: 16, label: "Strawberry", value: "strawberry" },
  { id: 17, label: "Tangerine", value: "tangerine" },
  { id: 18, label: "Ugli fruit", value: "ugli-fruit" },
  { id: 19, label: "Vanilla bean", value: "vanilla-bean" },
  { id: 20, label: "Watermelon", value: "watermelon" },
  { id: 21, label: "Blackberry", value: "blackberry" },
  { id: 22, label: "Blueberry", value: "blueberry" },
  { id: 23, label: "Cranberry", value: "cranberry" },
  { id: 24, label: "Grapefruit", value: "grapefruit" },
  { id: 25, label: "Pineapple", value: "pineapple" },
  { id: 26, label: "Pomegranate", value: "pomegranate" },
  { id: 27, label: "Apricot", value: "apricot" },
  { id: 28, label: "Avocado", value: "avocado" },
  { id: 29, label: "Coconut", value: "coconut" },
  { id: 30, label: "Dragonfruit", value: "dragonfruit" },
  { id: 31, label: "Guava", value: "guava" },
  { id: 32, label: "Jackfruit", value: "jackfruit" },
  { id: 33, label: "Lime", value: "lime" },
  { id: 34, label: "Lychee", value: "lychee" },
  { id: 35, label: "Nectarine", value: "nectarine" },
  { id: 36, label: "Passion fruit", value: "passion-fruit" },
  { id: 37, label: "Peach", value: "peach" },
  { id: 38, label: "Pear", value: "pear" },
  { id: 39, label: "Persimmon", value: "persimmon" },
  { id: 40, label: "Plum", value: "plum" },
];
