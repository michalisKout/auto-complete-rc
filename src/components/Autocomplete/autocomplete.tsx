import { useCallback, useState } from "react";
import "./Autocomplete.css";
import { useAccessibility } from "./hooks/useAccessibility";
import { useAutocomplete } from "./hooks/useAutocomplete";
import { useClickOutside } from "./hooks/useClickOutside";
import { useScrollSync } from "./hooks/useScrollSync";
import type { AutocompleteItem, AutocompleteProps } from "./types";
import { debounce } from "./utils/debounce";

/**
 * Autocomplete component for searching and selecting items
 *
 * @param {AutocompleteProps} props - Component properties
 * @returns {JSX.Element} Rendered Autocomplete component
 */
export const Autocomplete: React.FC<AutocompleteProps> = ({
  onInputChange,
  placeholder = "Search...",
  className = "",
  disabled = false,
  testId = "test-id",
  defaultItems = [],
  onSelect,
  filterItems,
  debounceMs = 500,
  minChars = 1,
  noResultsText = "No results found",
  loadingText = "Searching...",
}) => {
  const inputId = `autocomplete-input-${testId}`;
  const dropdownId = `autocomplete-dropdown-${testId}`;
  const [inputValue, setInputValue] = useState("");
  const [isOpen, setIsOpen] = useState(false);
  const [selectedItem, setSelectedItem] = useState<AutocompleteItem | null>(null);
  const { filteredItems, isLoading, asyncSearch, error } = useAutocomplete(filterItems, {
    defaultItems,
    minChars,
  });

  const { accessibilityProps, setFocusedIndex, focusedIndex, handleKeyDown } = useAccessibility({
    isOpen,
    testId: dropdownId,
    filteredItems,
    onEnter: (item) => {
      setSelectedItem(item);
      setInputValue(item.label);
      onInputChange?.(item.label);
      setIsOpen(false);
    },
    onReset: () => {
      setIsOpen(false);
    },
  });

  useScrollSync({
    focusedIndex,
    dropdownId,
    isOpen,
  });

  const { ref } = useClickOutside<HTMLDivElement>({
    handleClick: () => {
      setIsOpen(false);
    },
    isOpen,
  });

  const handleSearch = useCallback(
    async (value: string) => {
      setIsOpen(value.length >= minChars && !disabled);
      onInputChange?.(value);
      await asyncSearch(value);
    },
    [onInputChange, disabled, asyncSearch, minChars],
  );

  const debouncedSearch = useCallback(debounce(handleSearch, debounceMs), []);

  const handleInputChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    const value = event.target.value;
    setInputValue(value);
    debouncedSearch(value);
  };

  const handleItemSelect = useCallback(
    (item: AutocompleteItem) => {
      setInputValue(item.label);
      onInputChange?.(item.label);
      setSelectedItem(item);
      setIsOpen(false);
      setFocusedIndex(-1);
      onSelect?.(item);
    },
    [onSelect, onInputChange, setFocusedIndex],
  );

  const handleInputFocus = useCallback(() => {
    setIsOpen(inputValue.length >= minChars);
  }, [inputValue.length, minChars]);

  const isSelected = useCallback(
    (itemId: number | string) => selectedItem?.id === itemId,
    [selectedItem],
  );

  const noResults = filteredItems.length === 0 && !isLoading && inputValue.length > 0;
  const hasError = !isLoading && error;
  const shouldShowItems = !isLoading && !noResults && !error;

  return (
    <div className={`autocomplete ${className}`} ref={ref}>
      <input
        {...accessibilityProps}
        onFocus={handleInputFocus}
        data-testid={inputId}
        type="text"
        className={`autocomplete__input ${error ? "autocomplete__input--error" : ""}`}
        value={inputValue}
        onChange={handleInputChange}
        placeholder={placeholder}
        disabled={disabled}
        autoComplete="off"
        onKeyDown={handleKeyDown}
      />

      {isOpen && (
        <div
          data-testid={dropdownId}
          aria-expanded={true}
          className="autocomplete__dropdown"
          role="listbox"
          aria-label="Search results"
          tabIndex={-1}
        >
          {hasError && <div className="autocomplete__error">{error}</div>}
          {isLoading && (
            <div className="autocomplete__loading">
              <div className="autocomplete__spinner" />
              {loadingText}
            </div>
          )}
          {shouldShowItems &&
            filteredItems?.map((item, index) => (
              <div
                key={item.id}
                id={`${dropdownId}-item-${index}`}
                className={`autocomplete__item ${isSelected(item.id) ? "autocomplete__item--selected" : ""}`}
                aria-selected={index === focusedIndex}
                onClick={() => handleItemSelect(item)}
              >
                {item.label}
              </div>
            ))}

          {noResults && !error && <div className="autocomplete__no-results">{noResultsText}</div>}
        </div>
      )}
    </div>
  );
};
