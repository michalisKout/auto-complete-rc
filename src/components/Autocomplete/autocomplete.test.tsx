import { render, screen, waitFor } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { beforeEach, describe, expect, it, vi } from "vitest";
import { mockData } from "../../../mockData";
import { Autocomplete } from "./autocomplete";
import type { AutocompleteItem, AutocompleteProps } from "./autocomplete.types";

describe("Autocomplete", () => {
  const defaultProps: AutocompleteProps = {
    onInputChange: vi.fn(),
    defaultItems: [],
    filterItems: vi.fn(),
  };

  beforeEach(() => {
    vi.clearAllMocks();
  });

  describe("Rendering", () => {
    it("renders with default props", () => {
      render(<Autocomplete {...defaultProps} />);

      const input = screen.getByTestId("autocomplete-input-test-id");
      expect(input).toBeInTheDocument();
      expect(input).toHaveAttribute("placeholder", "Search...");
      expect(input).toHaveValue("");
      expect(input).not.toBeDisabled();
    });

    it("renders with custom placeholder", () => {
      const placeholder = "Type to search...";
      render(<Autocomplete {...defaultProps} placeholder={placeholder} />);

      const input = screen.getByTestId("autocomplete-input-test-id");
      expect(input).toHaveAttribute("placeholder", placeholder);
    });

    it("renders with custom className", () => {
      const className = "custom-autocomplete";
      render(<Autocomplete {...defaultProps} className={className} />);

      const container = screen.getByTestId("autocomplete-input-test-id").parentElement;
      expect(container).toHaveClass("autocomplete", className);
    });
  });

  describe("Disabled State", () => {
    it("does not accept input when disabled", async () => {
      const mockOnInputChange = vi.fn();
      const user = userEvent.setup();

      render(<Autocomplete {...defaultProps} disabled={true} onInputChange={mockOnInputChange} />);

      const input = screen.getByTestId("autocomplete-input-test-id");
      await user.type(input, "test");

      expect(input).toHaveValue("");
      expect(mockOnInputChange).not.toHaveBeenCalled();
    });
  });

  describe("Async Filtering", () => {
    it("calls filterItems function when user types more than minimum characters", async () => {
      const mockFilterItems = vi.fn().mockResolvedValue([]);
      const user = userEvent.setup();

      render(<Autocomplete {...defaultProps} filterItems={mockFilterItems} />);

      const input = screen.getByTestId("autocomplete-input-test-id");
      await user.type(input, "ap");

      await waitFor(() => {
        expect(mockFilterItems).toHaveBeenCalledWith("ap", expect.any(AbortSignal));
      });
    });

    it("displays filtered results from async filterItems function", async () => {
      const filteredResults: AutocompleteItem[] = [
        { id: 1, label: "Apple", value: "apple" },
        { id: 27, label: "Apricot", value: "apricot" },
      ];
      const mockFilterItems = vi.fn().mockResolvedValue(filteredResults);
      const user = userEvent.setup();

      render(<Autocomplete {...defaultProps} filterItems={mockFilterItems} />);

      const input = screen.getByTestId("autocomplete-input-test-id");
      await user.type(input, "ap");

      await waitFor(() => {
        expect(screen.getByText("Apple")).toBeInTheDocument();
        expect(screen.getByText("Apricot")).toBeInTheDocument();
      });

      const dropdown = screen.getByTestId("autocomplete-dropdown-test-id");
      expect(dropdown).toBeInTheDocument();
    });

    it("allows item selection from async filtered results", async () => {
      const filteredResults: AutocompleteItem[] = [
        { id: 1, label: "Apple", value: "apple" },
        { id: 27, label: "Apricot", value: "apricot" },
      ];
      const mockFilterItems = vi.fn().mockResolvedValue(filteredResults);
      const mockOnSelect = vi.fn();
      const user = userEvent.setup();

      render(
        <Autocomplete {...defaultProps} filterItems={mockFilterItems} onSelect={mockOnSelect} />,
      );

      const input = screen.getByTestId("autocomplete-input-test-id");
      await user.type(input, "ap");

      // Wait for results to appear
      await waitFor(() => {
        expect(screen.getByText("Apple")).toBeInTheDocument();
      });

      // Click on an item
      const appleItem = screen.getByText("Apple");
      await user.click(appleItem);

      expect(mockOnSelect).toHaveBeenCalledWith({
        id: 1,
        label: "Apple",
        value: "apple",
      });

      // Input should be updated with selected item
      expect(input).toHaveValue("Apple");

      // Dropdown should close
      expect(screen.queryByTestId("autocomplete-dropdown-test-id")).not.toBeInTheDocument();
    });

    it("works with synchronous filterItems function", async () => {
      const syncFilterItems = vi.fn().mockImplementation((query: string) => {
        return mockData.filter((item) => item.label.toLowerCase().includes(query.toLowerCase()));
      });
      const user = userEvent.setup();

      render(<Autocomplete {...defaultProps} filterItems={syncFilterItems} />);

      const input = screen.getByTestId("autocomplete-input-test-id");
      await user.type(input, "ap");

      await waitFor(() => {
        expect(syncFilterItems).toHaveBeenCalledWith("ap", expect.any(AbortSignal));
        expect(screen.getByText("Apple")).toBeInTheDocument();
        expect(screen.getByText("Apricot")).toBeInTheDocument();
        expect(screen.getByText("Grape")).toBeInTheDocument();
        expect(screen.getByText("Grapefruit")).toBeInTheDocument();
        expect(screen.getByText("Pineapple")).toBeInTheDocument();
      });
    });

    it("displays no results message when search returns empty results", async () => {
      const mockFilterItems = vi.fn().mockResolvedValue([]);
      const user = userEvent.setup();

      render(<Autocomplete {...defaultProps} filterItems={mockFilterItems} />);

      const input = screen.getByTestId("autocomplete-input-test-id");
      await user.type(input, "no results");

      await waitFor(() => {
        expect(mockFilterItems).toHaveBeenCalledWith("no results", expect.any(AbortSignal));
      });

      const dropdown = screen.getByTestId("autocomplete-dropdown-test-id");
      expect(dropdown).toBeInTheDocument();
      expect(dropdown).toContainElement(screen.getByText("No results found"));
    });

    it("displays error message when search fails", async () => {
      const errorMessage = "API request failed";
      const mockFilterItems = vi.fn().mockRejectedValue(new Error(errorMessage));
      const user = userEvent.setup();

      render(<Autocomplete {...defaultProps} filterItems={mockFilterItems} />);

      const input = screen.getByTestId("autocomplete-input-test-id");
      await user.type(input, "test");

      await waitFor(() => {
        expect(mockFilterItems).toHaveBeenCalledWith("test", expect.any(AbortSignal));
      });

      const dropdown = screen.getByTestId("autocomplete-dropdown-test-id");
      expect(dropdown).toBeInTheDocument();
      expect(dropdown).toContainElement(screen.getByText(errorMessage));
      expect(screen.queryByText('No results found for "test"')).not.toBeInTheDocument();
    });
  });

  describe("Accessibility", () => {
    it("has proper ARIA attributes", () => {
      render(<Autocomplete {...defaultProps} />);

      const input = screen.getByTestId("autocomplete-input-test-id");

      expect(input).toHaveAttribute("role", "combobox");
      expect(input).toHaveAttribute("aria-expanded", "false");
      expect(input).toHaveAttribute("aria-haspopup", "listbox");
      expect(input).toHaveAttribute("autoComplete", "off");
    });

    it("scrolls focused item into view when navigating with arrow keys", async () => {
      const mockScrollIntoView = vi.fn();
      Element.prototype.scrollIntoView = mockScrollIntoView;

      const items: AutocompleteItem[] = Array.from({ length: 10 }, (_, i) => ({
        id: i,
        label: `Item ${i}`,
        value: `item-${i}`,
      }));

      const mockFilterItems = vi.fn().mockResolvedValue(items);
      const user = userEvent.setup();

      render(<Autocomplete {...defaultProps} filterItems={mockFilterItems} />);

      const input = screen.getByTestId("autocomplete-input-test-id");
      await user.type(input, "test");

      await waitFor(() => {
        expect(screen.getByText("Item 0")).toBeInTheDocument();
      });

      await user.keyboard("{ArrowDown}");

      await waitFor(
        () => {
          expect(mockScrollIntoView).toHaveBeenCalledWith({
            behavior: "smooth",
            block: "nearest",
          });
        },
        { timeout: 100 },
      );

      vi.restoreAllMocks();
    });
  });
});
