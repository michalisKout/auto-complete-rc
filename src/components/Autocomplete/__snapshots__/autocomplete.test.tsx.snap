// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Autocomplete > Async Filtering > allows item selection from async filtered results 1`] = `
<div
  aria-label="Search results"
  class="autocomplete__dropdown"
  data-testid="autocomplete-dropdown-test-id"
  role="listbox"
  tabindex="-1"
>
  <div
    aria-selected="false"
    class="autocomplete__item "
    data-testid="autocomplete-dropdown-test-id-item-0"
    id="autocomplete-dropdown-test-id-item-0"
  >
    <span>
      <span
        class="autocomplete__highlighted"
      >
        Ap
      </span>
      <span>
        ple
      </span>
    </span>
  </div>
  <div
    aria-selected="false"
    class="autocomplete__item "
    data-testid="autocomplete-dropdown-test-id-item-1"
    id="autocomplete-dropdown-test-id-item-1"
  >
    <span>
      <span
        class="autocomplete__highlighted"
      >
        Ap
      </span>
      <span>
        ricot
      </span>
    </span>
  </div>
</div>
`;

exports[`Autocomplete > Async Filtering > works with synchronous filterItems function 1`] = `
<div
  aria-label="Search results"
  class="autocomplete__dropdown"
  data-testid="autocomplete-dropdown-test-id"
  role="listbox"
  tabindex="-1"
>
  <div
    aria-selected="false"
    class="autocomplete__item "
    data-testid="autocomplete-dropdown-test-id-item-0"
    id="autocomplete-dropdown-test-id-item-0"
  >
    <span>
      <span
        class="autocomplete__highlighted"
      >
        Ap
      </span>
      <span>
        ple
      </span>
    </span>
  </div>
  <div
    aria-selected="false"
    class="autocomplete__item "
    data-testid="autocomplete-dropdown-test-id-item-1"
    id="autocomplete-dropdown-test-id-item-1"
  >
    <span>
      <span>
        Gr
      </span>
      <span
        class="autocomplete__highlighted"
      >
        ap
      </span>
      <span>
        e
      </span>
    </span>
  </div>
  <div
    aria-selected="false"
    class="autocomplete__item "
    data-testid="autocomplete-dropdown-test-id-item-2"
    id="autocomplete-dropdown-test-id-item-2"
  >
    <span>
      <span>
        P
      </span>
      <span
        class="autocomplete__highlighted"
      >
        ap
      </span>
      <span>
        aya
      </span>
    </span>
  </div>
  <div
    aria-selected="false"
    class="autocomplete__item "
    data-testid="autocomplete-dropdown-test-id-item-3"
    id="autocomplete-dropdown-test-id-item-3"
  >
    <span>
      <span>
        Gr
      </span>
      <span
        class="autocomplete__highlighted"
      >
        ap
      </span>
      <span>
        efruit
      </span>
    </span>
  </div>
  <div
    aria-selected="false"
    class="autocomplete__item "
    data-testid="autocomplete-dropdown-test-id-item-4"
    id="autocomplete-dropdown-test-id-item-4"
  >
    <span>
      <span>
        Pine
      </span>
      <span
        class="autocomplete__highlighted"
      >
        ap
      </span>
      <span>
        ple
      </span>
    </span>
  </div>
  <div
    aria-selected="false"
    class="autocomplete__item "
    data-testid="autocomplete-dropdown-test-id-item-5"
    id="autocomplete-dropdown-test-id-item-5"
  >
    <span>
      <span
        class="autocomplete__highlighted"
      >
        Ap
      </span>
      <span>
        ricot
      </span>
    </span>
  </div>
</div>
`;
